modules/haze_removal/src/dark_channel_algorithm.cpp

```diff
@@ -0,0 +1,255 @@
+    {
+        const Mat src = _src.getMat();
+        if (src.channels() != 3)
+            CV_Error(Error::BadNumChannels,
+                     "Unsupported channels count: Only 3 channel BGR images supported");
+        erode(src, _dst, kernelForEroding);
+        std::vector<Mat> planes(3);
+        split(_dst, planes);
+        _dst.assign(min(planes[2], min(planes[1], planes[0])));
+        return;
+    }
+
+    Scalar getAtmoLight(cv::InputArray _src)
+    {
+
+        const Mat src = _src.getMat();
+        Mat darkChannel;
+        getDarkChannel(src, darkChannel);
+
+        int pixelsToKeep = src.rows * src.cols * percentageBrightestPixelsForAtmoLight;
+        int thresholdAtmo = intensityOfNthBrightestPixel(darkChannel, pixelsToKeep);
+
+        // Apply a threshold to get a mask of the $percentAtmo brightest pixels.
+        Mat dst;
+        threshold(darkChannel, dst, thresholdAtmo, 255, cv::THRESH_BINARY);
+
+        Scalar atmoLight = mean(src, dst);
+
+        return atmoLight;
+    }
+
+    void getTransmission(cv::InputArray _src, cv::OutputArray _dst, Scalar atmoLight)
+    {
+        const Mat src = _src.getMat();
+        Mat dst;
+        Mat src_float;
+        src.convertTo(src_float, CV_32FC3);
+        Mat src_normalized = src_float / atmoLight;
+        getDarkChannel(src_normalized, dst);
+        dst = 1 - dst * omega;
+        _dst.assign(dst);
+    }
+
+    void refineTransmission(cv::InputArray _src, cv::InputArray _unrefinedTransmission, cv::OutputArray _refinedTransmission)
+    {
+        Mat gray;
+        cvtColor(_src, gray, COLOR_BGR2GRAY);
+        ximgproc::guidedFilter(gray, _unrefinedTransmission, _refinedTransmission, guidedFilterRadius, guidedFilterEps);
+    }
+
+    virtual void dehaze(cv::InputArray _src, cv::OutputArray _dst) CV_OVERRIDE
+    {
+        const cv::Mat src = _src.getMat();
+        CV_Assert(src.channels() == 3);
+        CV_Assert(src.type() == CV_8UC3);
+
+        Scalar atmoLight = getAtmoLight(src);
+
+        Mat unrefinedTransmission;
+        getTransmission(src, unrefinedTransmission, atmoLight);
+
+        Mat refinedTransmission;
+
+        refineTransmission(src, unrefinedTransmission, refinedTransmission);
+        refinedTransmission = max(refinedTransmission, transmissionLowerBound);
+
+        std::vector<Mat> planes;
+
+        for (int i = 0; i < 3; i++)
+            planes.push_back(refinedTransmission);
+
+        merge(planes, refinedTransmission);
+
+        Mat src_32FC3;
+        src.convertTo(src_32FC3, CV_32F);
+
+        Mat radiance_32FC3;
+        Mat radiance;
+
+        radiance_32FC3 = (src_32FC3 - atmoLight) / refinedTransmission + atmoLight;
+        radiance_32FC3.convertTo(radiance, CV_8U);
+
+        _dst.assign(radiance);
+
+        return;
+    }
+};
+
+inline HazeRemovalHeImpl *getLocalImpl(HazeRemovalBase::HazeRemovalImpl *ptr)
+{
+    HazeRemovalHeImpl *impl = static_cast<HazeRemovalHeImpl *>(ptr);
+    CV_Assert(impl);
```

Comment:
or dynamic_cast ?

Reply:
It should have been dynamic_cast. Fixed that.